vowpalwabbit/oaa.cc

```diff
@@ -85,11 +88,94 @@ void predict_or_learn(oaa& o, LEARNER::base_learner& base, example& ec)
     o.all->print_text(o.all->raw_prediction, outputStringStream.str(), ec.tag);
   }
 
-  ec.pred.multiclass = prediction;
+  if (is_probabilities)
+  { float sum_prob = 0;
+    for (uint32_t i=0; i<o.k; i++)
+    { // probability of class (i+1) = logistic_link_function(raw_prediction)
+      float prob = 1.f / (1.f + exp(- o.pred[i].scalar));
+      o.probs[i] = prob;
+      sum_prob += prob;
+    }
+    // make sure that the probabilities sum up (exactly) to one
+    for (uint32_t i=0; i<o.k; i++)
+      o.probs[i] /= sum_prob;
+    // copy probs pointer to the example.
+    // TODO: shouldn't we do a deep copy and free(ec.pred.probs) in finish_example()?
+    ec.pred.probs = o.probs;
+  } else {
+    ec.pred.multiclass = prediction;
+  }
+
   ec.l.multi = mc_label_data;
 }
 
-void finish(oaa&o) { free(o.pred); free(o.subsample_order); }
+void finish(oaa&o)
+{ free(o.pred);
+  free(o.subsample_order);
+  free(o.probs);
+}
+
+// TODO: partial code duplication with multiclass.cc:finish_example
+void finish_example_probabilities(vw& all, oaa& o, example& ec)
+{ // === Compute multiclass_log_loss
+  // TODO:
+  // What to do if the correct label is unknown, i.e. (uint32_t)-1?
```

Comment:
In the final resume, I want to report both zero-one loss and multiclass log loss. So I need a separate field to store both.

BTW: I tried to understand the existing accounting mechanism, but it was quite difficult as it is scattered over several files. One of the purposes of this TODO note was to check also the *existing* mechanism for possible bugs (e.g. with regards to `all.sd->weighted_unlabeled_examples` handling and unknown labels).